### PR TITLE
feat(TCOMP-283): Add automatic type converter.

### PR DESCRIPTION
**Pull request by pbailly**
**Please check if the PR fulfills these requirements**

- [ ] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format) ?
- [ ] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [ ] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?**

- [ ] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/299/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/299/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/302/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/302/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/305/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/305/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/307/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/307/)

**Comment by gonchar-ivan** [comment]
I think these classes don't duplicate classes in org.talend.daikon.avro.converter.

avro.converter package contains classes for data(user-specific types)-to-avro and avro-to-data conversion

This container package contains classes for data-to-data conversion (conversion from some type to some other type, but not avro related). (If I understand correctly). Maybe it is something like data-to-talendType converters... I'm not sure.

My comments:
1) It should be covered by unit-tests more thoroughly
2) Please, provide javadocs at least to explain when this functionality should be used. Ideally, we need a documentation page about different data conversion (talend types to avro mapping; other types to avro mapping etc)

**Comment by pbailly** [comment]
Hello Ivan,
This implementation was here to backport the Studio's [BigDataParser](https://github.com/Talend/tbd-studio-ee/blob/master/main/plugins/org.talend.designer.routines.bigdata/resources/java/routines/system/BigDataParserUtils.java) functionalities into daikon.

This currently contains the Studio's type, but in the future, I would really love to improve the genericity in order to be able to add any types at run time. I wrote an architecture document to specify our needs on this track, but I'm waiting for some internal review. It should really help to clarify the big picture behind this PR.

However, I'm really concern about your comment on unit-tests. With more than 1000 lines of tests, for 700 lines of code, I thought I did my job. Maybe we disagree on how we should write JUNIT tests. Could you specify what are your expectation?

**Comment by gonchar-ivan** [comment]
I'm sorry. I was wrong about unit-tests.

Initially, I found only BigDecimalConverterTest and TypeConverterTest.
I thought that other classes are not covered (I expected IntergerConverterTest, LongConverterTest etc)

However, I rechecked with jacoco and found 89% coverage.
So, good job! Thank you!
And sorry for comment.

As I know generally we expect 75% unit-test coverage (not matter how much lines of tests ;-) )

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/198/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/198/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/201/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/201/)

File: daikon/src/test/java/org/talend/daikon/container/TypeConverterTest.java [hunk 1]
```diff
@@ -0,0 +1,1024 @@
...
+        assertEquals(input_Double, TypeConverter.asDouble().convert(input_int));
+        assertEquals(input_Double, TypeConverter.asDouble().convert(input_Integer));
+        assertEquals(input_Double, TypeConverter.asDouble().convert(input_long));
+        assertEquals(input_Double, TypeConverter.asDouble().convert(input_Long));
+        assertEquals(input_Double, TypeConverter.asDouble().convert(input_short));
+        assertEquals(input_Double, TypeConverter.asDouble().convert(input_Short));
+    }
+
+    @Test
+    public void test_asDouble_default() {
```
Comment (gonchar-ivan) [inline daikon/src/test/java/org/talend/daikon/container/TypeConverterTest.java]: Agree with codacy-bot

Comment (pbailly) [inline daikon/src/test/java/org/talend/daikon/container/TypeConverterTest.java]: We always used the syntaxe "nameOfTheMethodToTest_TestCase" to clarify the current usecase of the unit test.
This is the only exception of the "no underscore" rule for me.

I asked multiple times Sebastien if this was all right and if we can remove the Codacy warning for the testing part, he agreed.

**Comment by pbailly** [comment]
Hello Ivan,
Thanks for the review.
These methods are an upgrade from the utility methods available in the Studio.
They are considered as a root element to correctly interpret user's data in an uniform way among any Talend product. Therefore I do think this code has its place into Daikon. If not, I can move it to TCOMP.

"org.talend.daikon/data/converter" is a great name. I will move the package.

File: daikon/src/test/java/org/talend/daikon/container/TypeConverterTest.java (hunk 1 above)
Comment (gonchar-ivan) [inline daikon/src/test/java/org/talend/daikon/container/TypeConverterTest.java]: I don't mind if Seb agreed.
I'm ok about test method names. What about variable names?
There are a lot in BigDecimalConverterTest 
e.g. "input_byte"
I don't know if it is good

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/13/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/13/)

**Comment by gonchar-ivan** [comment]
You need to format each source file properly using 
https://github.com/Talend/tools/tree/master/tools-java-formatter/src/main/resources

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/14/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/14/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/15/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/15/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/16/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/16/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/daikon_pull-request_master/17/) [:x:](https://ci-common.datapwn.com/job/daikon_pull-request_master/17/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/18/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/18/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/21/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/21/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/24/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/24/)

**Pull request merged**